wallet: Add CoinGrinder coin selection algorithm

murchandamus opened the pull request:
***Please refer to the [topic on Delving Bitcoin](https://delvingbitcoin.org/t/gutterguard-and-coingrinder-simulation-results/279) discussing Gutter Guard/Coingrinder simulation results.***

Adds a coin selection algorithm that minimizes the weight of the input set while creating change.

Motivations
---

- At high feerates, using unnecessary inputs can significantly increase the fees
- Users are upset when fees are relatively large compared to the amount sent
- Some users struggle to maintain a sufficient count of UTXOs in their wallet

Approach
---

So far, Bitcoin Core has used a balanced approach to coin selection, where it will generate multiple input set candidates using various coin selection algorithms and pick the least wasteful among their results, but not explicitly minimize the input set weight. Under some circumstances, we _do_ want to minimize the weight of the input set. Sometimes changeless solutions require many or heavy inputs, and there is not always a changeless solution for Branch and Bound to find in the first place. This can cause expensive transactions unnecessarily. Given a wallet with sufficient funds, `CoinGrinder` will pick the minimal-waste input set for a transaction with a change output. The current implementation only runs `CoinGrinder` at feerates over 3×long-term-feerate-estimate (by default 30 ṩ/vB), which may be a decent compromise between our goal to reduce costs for the users, but still permit transactions at lower feerates to naturally reduce the wallet’s UTXO pool to curb bloat.

Trade-offs
---

Simulations for my thesis on coin selection ([see Section 6.3.2.1 [PDF]](https://murch.one/erhardt2016coinselection.pdf)) suggest that minimizing the input set for all transactions tends to grind a wallet’s UTXO pool to dust (pun intended): an approach selecting inputs per coin-age-priority (in effect similar to “largest first selection”) on average produced a UTXO pool with 15× the UTXO count as Bitcoin Core’s Knapsack-based Coin Selection then (in 2016). Therefore, I do not recommend running `CoinGrinder` under all circumstances, but only at extreme feerates or when we have another good reason to minimize the input set for other reasons. In the long-term, we should introduce additional metrics to score different input set candidates, e.g. on basis of their privacy and wallet health impact, to pick from all our coin selection results, but until then, we may want to limit use of `CoinGrinder` in other ways.

murchandamus committed:
coinselection: Add CoinGrinder algorithm

CoinGrinder is a DFS-based coin selection algorithm that
deterministically finds the input set with the lowest weight creating a
change output.

murchandamus committed:
opt: Tie-break UTXO sort by waste for BnB

Since we are searching for the minimal waste, we sort UTXOs with equal
effective value by ascending waste to be able to cut barren branches
earlier.

murchandamus committed:
doc: Document max_weight on BnB

Xekyo commented:
At this point, I’m mostly looking for Concept reviews.

It might also be interesting to run `CoinGrinder` when the fees of a transaction surpass some relative portion of the recipient outputs’ value. Any thoughts on that?

t-bast commented:
Thanks for exploring this, I think this is an interesting idea! I'll comment based on my specific usage of the wallet for lightning service providers, please bear in mind that this is a very specific use case that is quite different from regular users (see #24795 for early details about this).

My understanding of the current goals of coin selection are that it tries to find a good balance between minimizing fees and consolidating utxos. The latter actually helps the former: consolidating utxos is usually good for performance, and if you do it when fees are low, you will avoid paying more fees in the future when fees are higher because you'll be able to use less inputs in your future transactions.

In the case of a lightning service provider, we'd rather avoid consolidating utxos, and are ready to pay more fees to keep a large enough utxo set to satisfy our users' needs. The main reasons for that are that:

- we are using 0-conf (in a setup where the user is trusting us to not double-spend)
- we are using low fees for those transactions (it's ok to keep transactions unconfirmed until the mempool clears, this is how we can keep the cost low for users)
- the change outputs from most of those transactions are _unsafe_ because they may be double-spent by a commitment transaction (and thus generally unusable until the the transaction confirms, except when used for that same user)

That last point is a consequence of [splicing](https://github.com/lightning/bolts/pull/863), let me know if you want me to explain that point in more details.

If we run out of _safe_ utxos, we cannot onboard new users, which may require us to transfer funds from a cold wallet while making sure we're creating many utxos, at a time where the on-chain fees may be high. Our ideal wallet would maintain a large pool of utxos of various sizes, to ensure we're always ready to onboard new users or sell liquidity to existing users.

I understand that this is the opposite of what most end users want and can't be achieved by the same algorithm: that's why I would find it very useful to have various coin selection algorithms in bitcoin core, that are tuned to different profiles/scenarios. I like that this algorithm feels like something we could use (at all feerates) to keep a wallet state that makes more sense for us.

Xekyo commented:
Changes in last force-push: https://github.com/bitcoin/bitcoin/compare/10621c6c4499ba32ff3b7f0bd11b0351f36982b4..0446986f8dd87db29153618d380d159ec0016c1a

The approach has been simplified by keeping only track of the input weight instead of weight and waste, and minimizing input weight throughout. This reduced the number of parameters necessary to call CoinGrinder, simplified the input set traversal, and allows CoinGrinder to be used at all feerates instead of just high feerates. It is however still only called for feerates above 100 ṩ/vB from [`spend.cpp`](https://github.com/bitcoin/bitcoin/pull/27877/files#diff-6e06b309cd494ef5da4e78aa0929a980767edd12342137f268b9219167064d13R586) at this time.

dergoegge commented:
Ran the fuzzer for a bit:

```
fuzz: wallet/test/fuzz/coinselection.cpp:102: void wallet::coin_grinder_fuzz_target(FuzzBufferType): Assertion `result_knapsack->GetWeight() >= result_cg->GetWeight()' failed.
```

Input to reproduce:
```sh
$ echo "d3d3/wF383d3d3d3d3d3d3d3d3d5d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3d3fzd3d3d3d3d7BE
Wnd3d3d3d3V3d3d3d3d3d3l6d3d3d3d3d3f///8Fd3d3d3d3d3d3d3d3d1d3d3d3d3d3d3d3d3d3
d3d3d3d3d3d3d3d3d3d3d3d3Mnd3d3d3d3d5d3d3d3d3YS5hYWFhYWFhYWFhYWFhYWFhYWFhYUBh
YWEuAAAACmFhWWFhYWFhd3d3d3d3d3d3d3d3dzp3d3d3d3d3d3d3d3d3d3d3d/N3d3d3d3d3d3d3
d3d3eXd3d3d3d3d3d3d1d3d3d3d3d3d3dzJ3d3d3d3d3eXd3d3d3d3d3d3d1d3d3d3d3d3d3dzJ3
d3d3d3d3d3d3d3d3d3d3dyh3d3d3d3d3dxcC+xc=" | base64 -d > coingrinder_crash
```

dergoegge commented:
I would like to see if the CI catches the [crash](https://github.com/bitcoin/bitcoin/pull/27877#issuecomment-1598451117), now that #27919 has been merged. @Xekyo would you mind rebasing?

Xekyo commented:
Rebased on Master.

Changes since last rebase: SelectionResult now has a record if BnB or CoinGrinder run out of tries and thus have not found the optimal solution.

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -191,6 +191,161 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+    };
+
+    for (size_t curr_try = 0;;) {
+        if (curr_try >= TOTAL_TRIES || next_op == operations::done) {
```
No, I don’t think so, an equality check here should be functionally equivalent.

brunoerg commented on src/wallet/coinselection.cpp:
```diff
@@ -174,6 +174,10 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
         }
+        if (curr_try >= TOTAL_TRIES - 1) {
+            // On last attempt and didn’t break due to full traversal: solution is non-optimal
+            result.SetAlgoCompleted(false);
```
In cf14a72859a2eaf42a24a2ea3125bdadde5f6611: Here we're setting it for BnB but seems we're not using `GetAlgoCompleted` for it?

murchandamus replied:
You are right, I’ll have follow up on this

murchandamus reviewed (commented):
Rebased. Fixed up a couple comments and improved a commit message

murchandamus commented:
> The PR description should say that we don't use hardcoded fee rate value, but rather multiple of long-term-fee-rate. LTFR is an existing user configurable parameter. It's also a good fit for the purpose of determining when to be more "aggressive" with coin selection, because it captures forward looking fee market expectations.

Thanks, I’ve amended the description.

murchandamus reviewed (commented):
Rebased on #29037, addressed comments by @achow101 and @brunoerg

murchandamus commented:
Scenario: bustabit-tiny-hot-wallet_6h-YTD-2023

| Scenario File | Current Balance | Mean #UTXO | Current #UTXO | #Deposits | #Inputs Spent | #Withdraws | #Uneconomical outputs spent | #Change Created | #Changeless | Min Change Value | Max Change Value | Mean Change Value | Std. Dev. of Change Value | Total Fees | Mean Fees per Withdraw | Cost to Empty (10 sat/vB) | Total Cost | Min Input Size | Max Input Size | Mean Input Size | Std. Dev. of Input Size | Usage |
|  --- | --- | --- | --- | --- | ---| --- | --- | --- | --- |--- | --- | --- | --- | --- |--- | --- | --- | --- | --- | --- | --- | --- |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 0E-8 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | Total: **0** | 0 | 0 | 0 | 0 | 0 | 0 | 0E-66 | 0E-66 | 0 | 0 | 0 | 0 |  |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 15.19132132 | 41.5520000000 | 82 | 267 | 377 | 233 | 0 | 192 | bnb: **41** ; Total: **41** | 0.00056094 | 4.99693990 | 0.623982837969 | 1.35047741271 | 0.00753835 | 0.0000323534334764 | 0.000557600000 | 0.008095950000 | 1 | 13 | 1.61802575107 | 1.30813814098 | knapsack: **71** ; bnb: **41** ; srd: **96** ; cg: **25** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 23.91267006 | 95.9050000000 | 149 | 594 | 736 | 406 | 0 | 291 | bnb: **115** ; Total: **115** | 0.00050005 | 4.99693990 | 0.453797352131 | 1.14213144768 | 0.02265411 | 0.0000557983004926 | 0.00101320000 | 0.02366731000 | 1 | 16 | 1.81280788177 | 1.71109191837 | knapsack: **79** ; bnb: **115** ; srd: **126** ; cg: **86** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 27.26139099 | 139.078666667 | 272 | 961 | 1042 | 539 | 0 | 353 | bnb: **186** ; Total: **186** | 0.00050005 | 4.99693990 | 0.384915890538 | 1.05102526981 | 0.03321401 | 0.0000616215398887 | 0.00184960000 | 0.03506361000 | 1 | 16 | 1.93320964750 | 1.77319757552 | knapsack: **83** ; bnb: **186** ; srd: **148** ; cg: **122** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 19.63452421 | 190.328500000 | 445 | 1316 | 1304 | 684 | 0 | 433 | bnb: **251** ; Total: **251** | 0.00050005 | 8.99972093 | 0.372520674758 | 1.12084593688 | 0.04729994 | 0.0000691519590643 | 0.00302600000 | 0.05032594000 | 1 | 16 | 1.90643274854 | 1.72400657664 | knapsack: **86** ; bnb: **251** ; srd: **163** ; cg: **184** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 26.26152510 | 250.734800000 | 534 | 1682 | 1630 | 818 | 3 | 482 | bnb: **336** ; Total: **336** | 0.00050005 | 8.99972093 | 0.339629060207 | 1.06772167819 | 0.05283593 | 0.0000645916014670 | 0.00363120000 | 0.05646713000 | 1 | 16 | 1.99266503667 | 1.76736219929 | knapsack: **91** ; bnb: **336** ; srd: **182** ; cg: **209** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 18.33155379 | 280.650000000 | 282 | 2008 | 2291 | 992 | 3 | 565 | bnb: **427** ; Total: **427** | 0.00050005 | 8.99972093 | 0.296057988053 | 0.992420318575 | 0.07125368 | 0.0000718283064516 | 0.00191760000 | 0.07317128000 | 1 | 68 | 2.30947580645 | 3.62884382343 | knapsack: **103** ; bnb: **427** ; srd: **211** ; cg: **251** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 26.34484158 | 295.944000000 | 497 | 2415 | 2533 | 1085 | 3 | 615 | bnb: **470** ; Total: **470** | 0.00050005 | 8.99972093 | 0.301033089561 | 1.00396217082 | 0.08354896 | 0.0000770036497696 | 0.00337960000 | 0.08692856000 | 1 | 68 | 2.33456221198 | 3.54771818942 | knapsack: **106** ; bnb: **470** ; srd: **221** ; cg: **288** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 23.40685064 | 321.172750000 | 522 | 2818 | 2947 | 1182 | 3 | 651 | bnb: **531** ; Total: **531** | 0.00050005 | 8.99972093 | 0.287157889601 | 0.977740948544 | 0.09213523 | 0.0000779485871404 | 0.00354960000 | 0.09568483000 | 1 | 68 | 2.49323181049 | 3.76213203809 | knapsack: **111** ; bnb: **531** ; srd: **230** ; cg: **310** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 15.27722359 | 339.373555556 | 362 | 3213 | 3554 | 1287 | 3 | 703 | bnb: **584** ; Total: **584** | 0.00050005 | 8.99972093 | 0.271071994011 | 0.943740278451 | 0.10894035 | 0.0000846467365967 | 0.00246160000 | 0.11140195000 | 1 | 68 | 2.76146076146 | 4.27038778818 | knapsack: **124** ; bnb: **584** ; srd: **240** ; cg: **339** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 15.68972959 | 346.311400000 | 338 | 3571 | 4013 | 1429 | 3 | 780 | bnb: **649** ; Total: **649** | 0.00050005 | 8.99972093 | 0.260421651859 | 0.931312931704 | 0.12330692 | 0.0000862889573128 | 0.00229840000 | 0.12560532000 | 1 | 68 | 2.80825752274 | 4.51603605228 | knapsack: **128** ; bnb: **649** ; srd: **251** ; cg: **401** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 21.91471596 | 336.786000000 | 128 | 3938 | 4660 | 1562 | 3 | 850 | bnb: **712** ; Total: **712** | 0.00050005 | 8.99972093 | 0.288234402494 | 1.01494195839 | 0.13825175 | 0.0000885094430218 | 0.000870400000 | 0.139122150000 | 1 | 93 | 2.98335467350 | 5.35334241066 | knapsack: **139** ; bnb: **712** ; srd: **279** ; cg: **432** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 20.33460103 | 315.198000000 | 136 | 4278 | 5099 | 1722 | 3 | 957 | bnb: **765** ; Total: **765** | 0.00050005 | 8.99972093 | 0.297627968892 | 1.00386252310 | 0.14906261 | 0.0000865636527294 | 0.000924800000 | 0.149987410000 | 1 | 93 | 2.96109175377 | 5.16741191808 | knapsack: **156** ; bnb: **765** ; srd: **340** ; cg: **461** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 20.80829124 | 305.963230769 | 192 | 4650 | 5479 | 1850 | 3 | 1021 | bnb: **828** ; knapsack: **1** ; Total: **829** | 0.00050005 | 8.99972093 | 0.286942750392 | 0.974262801548 | 0.15671334 | 0.0000847099135135 | 0.00130560000 | 0.15801894000 | 1 | 93 | 2.96162162162 | 5.04089684085 | knapsack: **166** ; bnb: **828** ; srd: **362** ; cg: **494** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 27.15795622 | 297.497714286 | 128 | 4990 | 5981 | 2010 | 3 | 1119 | bnb: **890** ; knapsack: **1** ; Total: **891** | 0.00050005 | 8.99972093 | 0.268203352806 | 0.934307055742 | 0.17172147 | 0.0000854335671642 | 0.000870400000 | 0.172591870000 | 1 | 93 | 2.97562189055 | 4.93944442582 | knapsack: **186** ; bnb: **890** ; srd: **387** ; cg: **547** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 16.85821784 | 289.089333333 | 261 | 5335 | 6304 | 2165 | 3 | 1230 | bnb: **934** ; knapsack: **1** ; Total: **935** | 0.00050005 | 8.99972093 | 0.263149622317 | 0.912339284650 | 0.18921494 | 0.0000873972009238 | 0.00177480000 | 0.19098974000 | 1 | 93 | 2.91177829099 | 4.80652233495 | knapsack: **188** ; bnb: **934** ; srd: **432** ; cg: **611** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 12.77225916 | 277.434125000 | 52 | 5654 | 6939 | 2346 | 3 | 1337 | bnb: **1008** ; knapsack: **1** ; Total: **1009** | 0.00050005 | 8.99972093 | 0.293228898549 | 0.983960310643 | 0.19890860 | 0.0000847862745098 | 0.000353600000 | 0.199262200000 | 1 | 93 | 2.95780051151 | 4.89544508935 | knapsack: **207** ; bnb: **1008** ; srd: **501** ; cg: **630** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 26.46618688 | 265.148470588 | 111 | 5963 | 7343 | 2537 | 3 | 1491 | bnb: **1045** ; knapsack: **1** ; Total: **1046** | 0.00050005 | 8.99972093 | 0.288366364735 | 0.972127060772 | 0.21582423 | 0.0000850706464328 | 0.000754800000 | 0.216579030000 | 1 | 93 | 2.89436342136 | 4.76496850368 | knapsack: **238** ; bnb: **1045** ; srd: **559** ; cg: **695** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 15.44127335 | 260.461666667 | 181 | 6278 | 7698 | 2722 | 3 | 1601 | bnb: **1120** ; knapsack: **1** ; Total: **1121** | 0.00050005 | 8.99972093 | 0.272923193210 | 0.940972137535 | 0.23088255 | 0.0000848209221161 | 0.00123080000 | 0.23211335000 | 1 | 93 | 2.82806759735 | 4.62609823410 | knapsack: **239** ; bnb: **1120** ; srd: **593** ; cg: **770** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 19.39774696 | 254.855368421 | 196 | 6573 | 8122 | 2927 | 3 | 1745 | bnb: **1181** ; knapsack: **1** ; Total: **1182** | 0.00050005 | 8.99972093 | 0.258861722653 | 0.904744271874 | 0.24717869 | 0.0000844477929621 | 0.00133280000 | 0.24851149000 | 1 | 93 | 2.77485480014 | 4.49563581622 | knapsack: **262** ; bnb: **1181** ; srd: **659** ; cg: **825** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 13.24350003 | 249.937900000 | 141 | 6892 | 8610 | 3108 | 3 | 1859 | bnb: **1248** ; knapsack: **1** ; Total: **1249** | 0.00050005 | 8.99972093 | 0.250384759333 | 0.881788647427 | 0.25638219 | 0.0000824910521236 | 0.000958800000 | 0.257340990000 | 1 | 93 | 2.77027027027 | 4.42008658560 | knapsack: **293** ; bnb: **1248** ; srd: **716** ; cg: **851** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 12.86220801 | 243.836571429 | 50 | 7193 | 9163 | 3307 | 3 | 2020 | bnb: **1286** ; knapsack: **1** ; Total: **1287** | 0.00050005 | 9.49925129 | 0.308799488535 | 1.05657485993 | 0.28165986 | 0.0000851708073783 | 0.000340000000 | 0.281999860000 | 1 | 93 | 2.77078923496 | 4.65228505935 | knapsack: **324** ; bnb: **1286** ; srd: **775** ; cg: **922** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 11.98903214 | 236.785818182 | 91 | 7533 | 9567 | 3467 | 3 | 2125 | bnb: **1341** ; knapsack: **1** ; Total: **1342** | 0.00050005 | 9.49925129 | 0.306216955144 | 1.04354401456 | 0.29499272 | 0.0000850858725123 | 0.000618800000 | 0.295611520000 | 1 | 93 | 2.75944620710 | 4.58117079281 | knapsack: **337** ; bnb: **1341** ; srd: **818** ; cg: **971** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 27.04496050 | 232.355304348 | 259 | 7877 | 9866 | 3623 | 3 | 2248 | bnb: **1374** ; knapsack: **1** ; Total: **1375** | 0.00050005 | 9.49925129 | 0.304456446530 | 1.03785627472 | 0.30780143 | 0.0000849576124758 | 0.00176120000 | 0.30956263000 | 1 | 93 | 2.72315760420 | 4.50269344134 | knapsack: **353** ; bnb: **1374** ; srd: **878** ; cg: **1018** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 15.19652615 | 232.517333333 | 124 | 8204 | 10416 | 3796 | 3 | 2336 | bnb: **1459** ; knapsack: **1** ; Total: **1460** | 0.00050005 | 9.49925129 | 0.302783163489 | 1.03419376135 | 0.31692744 | 0.0000834898419389 | 0.000843200000 | 0.317770640000 | 1 | 93 | 2.74394099052 | 4.55304298606 | knapsack: **365** ; bnb: **1459** ; srd: **911** ; cg: **1061** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 24.28376881 | 230.030880000 | 157 | 8541 | 10825 | 3959 | 3 | 2441 | bnb: **1517** ; knapsack: **1** ; Total: **1518** | 0.00050005 | 9.49925129 | 0.294409770459 | 1.01446013114 | 0.32915702 | 0.0000831414549129 | 0.00106760000 | 0.33022462000 | 1 | 93 | 2.73427633241 | 4.50485277337 | knapsack: **387** ; bnb: **1517** ; srd: **954** ; cg: **1101** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 20.57080787 | 230.536384615 | 345 | 8875 | 11091 | 4125 | 3 | 2561 | bnb: **1563** ; knapsack: **1** ; Total: **1564** | 0.00050005 | 9.49925129 | 0.285567840449 | 0.993496387579 | 0.34327712 | 0.0000832186957576 | 0.00234600000 | 0.34562312000 | 1 | 93 | 2.68872727273 | 4.42439247168 | knapsack: **391** ; bnb: **1563** ; srd: **994** ; cg: **1177** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 14.63766153 | 230.698888889 | 177 | 9156 | 11656 | 4344 | 3 | 2677 | bnb: **1666** ; knapsack: **1** ; Total: **1667** | 0.00050005 | 9.49925129 | 0.279953263496 | 0.974355036268 | 0.35167860 | 0.0000809573204420 | 0.00120360000 | 0.35288220000 | 1 | 93 | 2.68324125230 | 4.39627403738 | knapsack: **422** ; bnb: **1666** ; srd: **1064** ; cg: **1192** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 21.30922579 | 229.831785714 | 207 | 9452 | 12063 | 4548 | 3 | 2818 | bnb: **1729** ; knapsack: **1** ; Total: **1730** | 0.00050005 | 9.49925129 | 0.269480742956 | 0.951948003836 | 0.36597978 | 0.0000804704881266 | 0.00140760000 | 0.36738738000 | 1 | 93 | 2.65237467018 | 4.31346530339 | knapsack: **473** ; bnb: **1729** ; srd: **1109** ; cg: **1237** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 17.48259680 | 231.010275862 | 243 | 9756 | 12445 | 4744 | 3 | 2932 | bnb: **1811** ; knapsack: **1** ; Total: **1812** | 0.00050005 | 9.49925129 | 0.260890875849 | 0.935058393028 | 0.38589250 | 0.0000813432757167 | 0.00165240000 | 0.38754490000 | 1 | 93 | 2.62331365936 | 4.24040039017 | knapsack: **477** ; bnb: **1811** ; srd: **1138** ; cg: **1318** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 17.79298599 | 230.726133333 | 276 | 10038 | 12849 | 4962 | 3 | 3087 | bnb: **1874** ; knapsack: **1** ; Total: **1875** | 0.00050005 | 9.49925129 | 0.251013131079 | 0.912751515855 | 0.40172562 | 0.0000809604232164 | 0.00187680000 | 0.40360242000 | 1 | 93 | 2.58948004837 | 4.16236608733 | knapsack: **501** ; bnb: **1874** ; srd: **1205** ; cg: **1382** |
| bustabit-tiny-hot-wallet_6h-YTD-2023 | 14.73253391 | 230.678403289 | 188 | 10076 | 12998 | 5005 | 3 | 3110 | bnb: **1894** ; knapsack: **1** ; Total: **1895** | 0.00050005 | 9.49925129 | 0.249207142897 | 0.909610046449 | 0.40248296 | 0.0000804161758242 | 0.00127840000 | 0.40376136000 | 1 | 93 | 2.59700299700 | 4.15295477730 | knapsack: **520** ; bnb: **1894** ; srd: **1209** ; cg: **1382** |

murchandamus commented:
Scenario: bustabit-tiny-hot-wallet_peak-and-tail

| Scenario File | Current Balance | Mean #UTXO | Current #UTXO | #Deposits | #Inputs Spent | #Withdraws | #Uneconomical outputs spent | #Change Created | #Changeless | Min Change Value | Max Change Value | Mean Change Value | Std. Dev. of Change Value | Total Fees | Mean Fees per Withdraw | Cost to Empty (10 sat/vB) | Total Cost | Min Input Size | Max Input Size | Mean Input Size | Std. Dev. of Input Size | Usage |
| bustabit-tiny-hot-wallet_peak-and-tail | 0E-8 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | Total: **0** | 0 | 0 | 0 | 0 | 0 | 0 | 0E-66 | 0E-66 | 0 | 0 | 0 | 0 |  |
|  --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| bustabit-tiny-hot-wallet_peak-and-tail | 15.19431007 | 23.8480000000 | 22 | 267 | 464 | 233 | 2 | 219 | bnb: **14** ; Total: **14** | 0.00051685 | 4.99597564 | 0.168985573607 | 0.679673163816 | 0.00454960 | 0.0000195261802575 | 0.000149600000 | 0.004699200000 | 1 | 13 | 1.99141630901 | 1.67123590523 | cg: **21** ; knapsack: **165** ; srd: **33** ; bnb: **14** |
| bustabit-tiny-hot-wallet_peak-and-tail | 23.92726978 | 29.1980000000 | 33 | 594 | 925 | 406 | 2 | 364 | bnb: **42** ; Total: **42** | 0.00050308 | 4.99597564 | 0.144475909643 | 0.610946105027 | 0.00805439 | 0.0000198383990148 | 0.000224400000 | 0.008278790000 | 1 | 15 | 2.27832512315 | 2.16797639109 | cg: **35** ; knapsack: **282** ; srd: **47** ; bnb: **42** |
| bustabit-tiny-hot-wallet_peak-and-tail | 27.28307178 | 34.2546666667 | 79 | 961 | 1356 | 539 | 3 | 474 | bnb: **65** ; Total: **65** | 0.00050188 | 4.99597564 | 0.128045802384 | 0.556617469511 | 0.01153322 | 0.0000213974397032 | 0.000537200000 | 0.012070420000 | 1 | 16 | 2.51576994434 | 2.58412941323 | cg: **57** ; knapsack: **356** ; srd: **61** ; bnb: **65** |
| bustabit-tiny-hot-wallet_peak-and-tail | 19.66739756 | 35.8575000000 | 14 | 1316 | 1899 | 684 | 5 | 597 | bnb: **87** ; Total: **87** | 0.00050188 | 9.84521291 | 0.183721836801 | 0.963651079172 | 0.01442659 | 0.0000210915058480 | 0.0000952000000 | 0.0145217900000 | 1 | 56 | 2.77631578947 | 3.52169106104 | cg: **67** ; knapsack: **453** ; srd: **77** ; bnb: **87** |
| bustabit-tiny-hot-wallet_peak-and-tail | 26.29724434 | 35.7344000000 | 51 | 1682 | 2340 | 818 | 5 | 709 | bnb: **109** ; Total: **109** | 0.00050188 | 9.84521291 | 0.200968485233 | 1.04753105765 | 0.01711669 | 0.0000209250488998 | 0.000346800000 | 0.017463490000 | 1 | 56 | 2.86063569682 | 3.52298146055 | cg: **77** ; knapsack: **547** ; srd: **85** ; bnb: **109** |
| bustabit-tiny-hot-wallet_peak-and-tail | 18.38145545 | 35.9136666667 | 29 | 2008 | 2841 | 992 | 5 | 862 | bnb: **130** ; Total: **130** | 0.00050188 | 9.84521291 | 0.227307114130 | 1.12133776363 | 0.02135202 | 0.0000215242137097 | 0.000197200000 | 0.021549220000 | 1 | 56 | 2.86391129032 | 3.41342810781 | cg: **98** ; knapsack: **652** ; srd: **112** ; bnb: **130** |
| bustabit-tiny-hot-wallet_peak-and-tail | 26.40480015 | 37.4480000000 | 48 | 2415 | 3307 | 1085 | 5 | 940 | bnb: **145** ; Total: **145** | 0.00050188 | 9.84521291 | 0.220479130457 | 1.08579016177 | 0.02359039 | 0.0000217422949309 | 0.000326400000 | 0.023916790000 | 1 | 56 | 3.04792626728 | 3.77717800053 | cg: **103** ; knapsack: **712** ; srd: **125** ; bnb: **145** |
| bustabit-tiny-hot-wallet_peak-and-tail | 23.47260037 | 38.7750000000 | 42 | 2818 | 3798 | 1182 | 5 | 1022 | bnb: **160** ; Total: **160** | 0.00050188 | 9.84521291 | 0.206872748151 | 1.04348216983 | 0.02638550 | 0.0000223227580372 | 0.000285600000 | 0.026671100000 | 1 | 56 | 3.21319796954 | 3.96686894231 | cg: **107** ; knapsack: **784** ; srd: **131** ; bnb: **160** |
| bustabit-tiny-hot-wallet_peak-and-tail | 15.35615672 | 39.6480000000 | 22 | 3213 | 4300 | 1287 | 5 | 1109 | bnb: **178** ; Total: **178** | 0.00050188 | 9.84521291 | 0.203870829522 | 1.00559062564 | 0.03000722 | 0.0000233156332556 | 0.000149600000 | 0.030156820000 | 1 | 56 | 3.34110334110 | 4.14898061580 | cg: **116** ; knapsack: **843** ; srd: **150** ; bnb: **178** |
| bustabit-tiny-hot-wallet_peak-and-tail | 15.78017875 | 39.6076000000 | 34 | 3571 | 4772 | 1429 | 5 | 1235 | bnb: **194** ; Total: **194** | 0.00050188 | 9.84521291 | 0.193004885474 | 0.955815690739 | 0.03285776 | 0.0000229935339398 | 0.000231200000 | 0.033088960000 | 1 | 56 | 3.33939818055 | 4.10168304958 | cg: **127** ; knapsack: **944** ; srd: **164** ; bnb: **194** |
| bustabit-tiny-hot-wallet_peak-and-tail | 22.01728937 | 38.6205454545 | 44 | 3938 | 5249 | 1562 | 5 | 1355 | bnb: **207** ; Total: **207** | 0.00050188 | 9.84521291 | 0.195981470052 | 0.951182116078 | 0.03567834 | 0.0000228414468630 | 0.000299200000 | 0.035977540000 | 1 | 56 | 3.36043533931 | 4.15167856608 | cg: **139** ; knapsack: **1045** ; srd: **171** ; bnb: **207** |
| bustabit-tiny-hot-wallet_peak-and-tail | 20.44464388 | 37.8906666667 | 27 | 4278 | 5747 | 1722 | 5 | 1496 | bnb: **226** ; Total: **226** | 0.00050188 | 9.84521291 | 0.188362870481 | 0.911149689949 | 0.03901976 | 0.0000226595586527 | 0.000183600000 | 0.039203360000 | 1 | 56 | 3.33739837398 | 4.04556927563 | cg: **150** ; knapsack: **1157** ; srd: **189** ; bnb: **226** |
| bustabit-tiny-hot-wallet_peak-and-tail | 20.92316834 | 37.9698461538 | 38 | 4650 | 6215 | 1850 | 5 | 1603 | bnb: **247** ; Total: **247** | 0.00050188 | 9.84521291 | 0.183975313481 | 0.883705979157 | 0.04183624 | 0.0000226141837838 | 0.000258400000 | 0.042094640000 | 1 | 56 | 3.35945945946 | 4.06278200898 | cg: **161** ; knapsack: **1242** ; srd: **200** ; bnb: **247** |
| bustabit-tiny-hot-wallet_peak-and-tail | 27.28430041 | 37.8664285714 | 72 | 4990 | 6668 | 2010 | 5 | 1750 | bnb: **260** ; Total: **260** | 0.00050188 | 9.84521291 | 0.177473687057 | 0.848859219835 | 0.04537728 | 0.0000225757611940 | 0.000489600000 | 0.045866880000 | 1 | 56 | 3.31741293532 | 4.01648485569 | cg: **174** ; knapsack: **1359** ; srd: **217** ; bnb: **260** |
| bustabit-tiny-hot-wallet_peak-and-tail | 16.99924857 | 37.4724000000 | 24 | 5335 | 7196 | 2165 | 5 | 1885 | bnb: **280** ; Total: **280** | 0.00050188 | 9.84521291 | 0.177391064403 | 0.843234149889 | 0.04818421 | 0.0000222559861432 | 0.000163200000 | 0.048347410000 | 1 | 56 | 3.32378752887 | 3.98576631937 | cg: **178** ; knapsack: **1472** ; srd: **235** ; bnb: **280** |
| bustabit-tiny-hot-wallet_peak-and-tail | 12.91955473 | 36.4260000000 | 23 | 5654 | 7687 | 2346 | 5 | 2056 | bnb: **290** ; Total: **290** | 0.00050188 | 9.84521291 | 0.193400275963 | 0.876458633010 | 0.05161303 | 0.0000220004390452 | 0.000156400000 | 0.051769430000 | 1 | 56 | 3.27664109122 | 3.90622814334 | cg: **189** ; knapsack: **1605** ; srd: **262** ; bnb: **290** |
| bustabit-tiny-hot-wallet_peak-and-tail | 26.62606905 | 36.0403529412 | 35 | 5963 | 8164 | 2537 | 5 | 2236 | bnb: **301** ; Total: **301** | 0.00050188 | 9.84521291 | 0.192812385640 | 0.870479296081 | 0.05594206 | 0.0000220504769413 | 0.000238000000 | 0.056180060000 | 1 | 56 | 3.21797398502 | 3.82928593235 | cg: **210** ; knapsack: **1742** ; srd: **284** ; bnb: **301** |
| bustabit-tiny-hot-wallet_peak-and-tail | 15.61256262 | 36.0920000000 | 28 | 6278 | 8656 | 2722 | 5 | 2406 | bnb: **316** ; Total: **316** | 0.00050188 | 9.84521291 | 0.184131431850 | 0.841978063836 | 0.05959328 | 0.0000218931961793 | 0.000190400000 | 0.059783680000 | 1 | 56 | 3.18001469508 | 3.77099867645 | cg: **223** ; knapsack: **1883** ; srd: **300** ; bnb: **316** |
| bustabit-tiny-hot-wallet_peak-and-tail | 19.58161842 | 35.7968421053 | 35 | 6573 | 9138 | 2927 | 5 | 2600 | bnb: **327** ; Total: **327** | 0.00050188 | 9.84521291 | 0.177131851981 | 0.813559033907 | 0.06330723 | 0.0000216287085753 | 0.000238000000 | 0.063545230000 | 1 | 56 | 3.12196788521 | 3.68354571639 | cg: **241** ; knapsack: **2041** ; srd: **318** ; bnb: **327** |
| bustabit-tiny-hot-wallet_peak-and-tail | 13.43288840 | 35.8227000000 | 26 | 6892 | 9630 | 3108 | 5 | 2764 | bnb: **344** ; Total: **344** | 0.00050188 | 9.84521291 | 0.175901789258 | 0.794930663393 | 0.06699382 | 0.0000215552831403 | 0.000176800000 | 0.067170620000 | 1 | 56 | 3.09845559846 | 3.63899917037 | cg: **254** ; knapsack: **2165** ; srd: **345** ; bnb: **344** |
| bustabit-tiny-hot-wallet_peak-and-tail | 13.07353548 | 35.0930476190 | 41 | 7193 | 10109 | 3307 | 5 | 2957 | bnb: **350** ; Total: **350** | 0.00050188 | 9.84521291 | 0.199681650183 | 0.868166694510 | 0.07033239 | 0.0000212677320835 | 0.000278800000 | 0.070611190000 | 1 | 56 | 3.05684910795 | 3.58448766707 | cg: **264** ; knapsack: **2321** ; srd: **372** ; bnb: **350** |
| bustabit-tiny-hot-wallet_peak-and-tail | 12.20992629 | 34.6953636364 | 42 | 7533 | 10598 | 3467 | 5 | 3107 | bnb: **360** ; Total: **360** | 0.00050188 | 9.84521291 | 0.195061106501 | 0.850538636883 | 0.07409857 | 0.0000213725324488 | 0.000285600000 | 0.074384170000 | 1 | 56 | 3.05682145948 | 3.55955862963 | cg: **282** ; knapsack: **2436** ; srd: **389** ; bnb: **360** |
| bustabit-tiny-hot-wallet_peak-and-tail | 27.27557467 | 34.8666956522 | 46 | 7877 | 11073 | 3623 | 5 | 3242 | bnb: **381** ; Total: **381** | 0.00050188 | 9.84521291 | 0.194723731783 | 0.845497044809 | 0.07718726 | 0.0000213047916092 | 0.000312800000 | 0.077500060000 | 1 | 56 | 3.05630692796 | 3.54308159924 | cg: **291** ; knapsack: **2543** ; srd: **408** ; bnb: **381** |
| bustabit-tiny-hot-wallet_peak-and-tail | 15.43327304 | 34.7681666667 | 22 | 8204 | 11580 | 3796 | 5 | 3398 | bnb: **398** ; Total: **398** | 0.00050188 | 9.84521291 | 0.194726944123 | 0.838401581678 | 0.08018055 | 0.0000211223788198 | 0.000149600000 | 0.080330150000 | 1 | 56 | 3.05057955743 | 3.51987725580 | cg: **297** ; knapsack: **2671** ; srd: **430** ; bnb: **398** |
| bustabit-tiny-hot-wallet_peak-and-tail | 24.52869452 | 34.8472000000 | 45 | 8541 | 12044 | 3959 | 5 | 3548 | bnb: **411** ; Total: **411** | 0.00050188 | 9.84521291 | 0.191581752049 | 0.824176340524 | 0.08423131 | 0.0000212759055317 | 0.000306000000 | 0.084537310000 | 1 | 56 | 3.04218236929 | 3.49546575364 | cg: **313** ; knapsack: **2786** ; srd: **449** ; bnb: **411** |
| bustabit-tiny-hot-wallet_peak-and-tail | 20.82722998 | 34.8699230769 | 28 | 8875 | 12548 | 4125 | 5 | 3701 | bnb: **424** ; Total: **424** | 0.00050188 | 9.84521291 | 0.187581044696 | 0.810098196171 | 0.08685501 | 0.00002105576 | 0.000190400000 | 0.087045410000 | 1 | 56 | 3.04193939394 | 3.47283152463 | cg: **319** ; knapsack: **2922** ; srd: **460** ; bnb: **424** |
| bustabit-tiny-hot-wallet_peak-and-tail | 14.89897305 | 34.6282962963 | 19 | 9156 | 13050 | 4344 | 5 | 3913 | bnb: **431** ; Total: **431** | 0.00050188 | 9.84521291 | 0.183598327498 | 0.792459161390 | 0.09036708 | 0.0000208027348066 | 0.000129200000 | 0.090496280000 | 1 | 56 | 3.00414364641 | 3.41597355038 | cg: **331** ; knapsack: **3107** ; srd: **475** ; bnb: **431** |
| bustabit-tiny-hot-wallet_peak-and-tail | 21.58068710 | 34.7263571429 | 25 | 9452 | 13527 | 4548 | 6 | 4100 | bnb: **448** ; Total: **448** | 0.00050188 | 9.84521291 | 0.182674596534 | 0.782311884138 | 0.09451847 | 0.0000207824252419 | 0.000170000000 | 0.094688470000 | 1 | 56 | 2.97427440633 | 3.38283712343 | cg: **353** ; knapsack: **3250** ; srd: **497** ; bnb: **448** |
| bustabit-tiny-hot-wallet_peak-and-tail | 17.77059990 | 34.5476551724 | 28 | 9756 | 14017 | 4744 | 6 | 4289 | bnb: **455** ; Total: **455** | 0.00050188 | 9.84521291 | 0.180617859419 | 0.771229915193 | 0.09788940 | 0.0000206343591906 | 0.000190400000 | 0.098079800000 | 1 | 56 | 2.95467959528 | 3.34763711902 | cg: **363** ; knapsack: **3409** ; srd: **517** ; bnb: **455** |
| bustabit-tiny-hot-wallet_peak-and-tail | 18.09223050 | 34.3252666667 | 36 | 10038 | 14499 | 4962 | 6 | 4497 | bnb: **465** ; Total: **465** | 0.00050188 | 9.84521291 | 0.178702173851 | 0.758088445077 | 0.10248111 | 0.0000206531862152 | 0.000244800000 | 0.102725910000 | 1 | 56 | 2.92200725514 | 3.30919619261 | cg: **388** ; knapsack: **3566** ; srd: **543** ; bnb: **465** |
| bustabit-tiny-hot-wallet_peak-and-tail | 15.03174881 | 34.3089317685 | 23 | 10076 | 14591 | 5005 | 6 | 4538 | bnb: **467** ; Total: **467** | 0.00050188 | 9.84521291 | 0.178563130670 | 0.755523891709 | 0.10326806 | 0.0000206329790210 | 0.000156400000 | 0.103424460000 | 1 | 56 | 2.91528471528 | 3.31017376335 | cg: **391** ; knapsack: **3593** ; srd: **554** ; bnb: **467** |

kashifs commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,156 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+ * • SHIFT: 010100 ⇒ 010010
+ *          the 4th is deselected and the 5th is selected instead
+ * • CUT:   010100 ⇒ 001000
+ *          the third is deselected, the first is deselected and the second is selected
```
For consistency, I think this should read:

`the 4th is deselected, the 2nd is deselected and the 3rd is selected`

murchandamus replied:
Thanks, I’ll fix that

kashifs commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,156 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+ * • CUT:   010100 ⇒ 001000
+ *          the third is deselected, the first is deselected and the second is selected
+ *
+ * @param const std::vector<OutputGroup>& utxo_pool The UTXO groups that we are choosing from.
```
This should read:

`@param std::vector<OutputGroup>& utxo_pool The UTXO groups that we are choosing from.`

murchandamus replied:
Fixed, thanks

kashifs commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,156 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+ *        value, with lower weight preferred as a tie-breaker.
+ * @param const CAmount& selection_target This is the minimum amount that we need for the transaction without
+ *        considering change.
+ * @param CAmount& change_target The minimum budget for creating a change output that we add to the selection_target.
```
this should read:

`@param CAmount change_target The minimum budget for creating a change output that we add to the selection_target.`

brunoerg replied:
Perhaps the comment is not wrong, the parameter should be updated. Doesn't it?

brunoerg replied:
perhaps `change_target` could be `const CAmount&`?

kashifs replied:
Perhaps. That would make it consistent with const CAmount& selection_target. It's my understanding that the documentation should exactly match the function definition. Is there another way to look at this?

murchandamus replied:
Thanks, I’ve amended the documentation

kashifs commented:
concept ACK [ef8ed8](https://github.com/bitcoin/bitcoin/commit/ef8ed8c973b2a48ccff54e4443a88e876d210bcf)

murchandamus reviewed (commented):
Rebased and addressed comments by @kashifs. Thanks for the thorough read, @kashifs.

murchandamus commented:
In a [simulation](https://delvingbitcoin.org/t/gutterguard-and-coingrinder-simulation-results/279) comparing the `master` branch with the `CoinGrinder` branch playing through a scenario that performs about 5 000 payments with feerates sampled from 2023, ___CoinGrinder reduces the total transaction fees by 9.5%___.

murchandamus commented:
After discussing Pieter’s review comments further with him, I decided to restructure the algorithm proposed in this PR. I’m returning the PR to draft for a few days while I rewrite the commits.

murchandamus committed:
opt: Skip over barren combinations of tiny UTXOs

Given a lot of small amount UTXOs it is possible that the lookahead
indicates sufficient funds, but any combination of them would push us
beyond the current best_weight.
We can estimate a lower bound for the minimal necessary weight to reach
target from the maximal amount and minimal weight in the tail of the
UTXO pool: if adding a number of hypothetical UTXOs of this maximum
amount and minimum weight would not be able to beat `best_weight`, we
can SHIFT to the omission branch, and CUT if the last selected UTXO is
not heavier than the minimum weight of the remainder.

murchandamus committed:
opt: Skip checking max_weight separately

Initialize `best_selection_weight` as `max_weight` allows us to skip the
separate `max_weight` check on every loop.

murchandamus committed:
opt: Cut if last addition was minimal weight

In situations where we have UTXO groups of various weight, we can CUT
rather than SHIFT when we exceeded the max_weight or the best
selection’s weight while the last step was equal to the minimum weight
in the lookahead.

murchandamus committed:
opt: Tiebreak UTXOs by weight for CoinGrinder

murchandamus committed:
opt: Skip evaluation of equivalent input sets

When two successive UTXOs match in effective value and weight, we can
skip the second if the prior is not selected: adding it would create an
equivalent input set to a previously evaluated.

E.g. if we have three UTXOs with effective values {5, 3, 3} of the same
weight each, we want to evaluate
{5, _, _}, {5, 3, _}, {5, 3, 3}, {_, 3, _}, {_, 3, 3},
but skip {5, _, 3}, and {_, _, 3}, because the first 3 is not selected,
and we therefore do not need to evaluate the second 3 at the same
position in the input set.

If we reach the end of the branch, we must SHIFT the previously selected
UTXO group instead.

murchandamus committed:
opt: Track remaining effective_value in lookahead

Introduces a dedicated data structure to track the total
effective_value available in the remaining UTXOs at each index of the
UTXO pool. In contrast to the approach in BnB, this allows us to
immediately jump to a lower index instead of visiting every UTXO to add
back their eff_value to the lookahead.

murchandamus committed:
opt: Skip branches with worse weight

Once we exceed the weight of the current best selection, we can always
shift as adding more inputs can never yield a better solution.

murchandamus committed:
fuzz: Add CoinGrinder fuzz target

murchandamus committed:
coinselection: Track whether CG completed

CoinGrinder may not be able to exhaustively search all potentially
interesting combinations for large UTXO pools, so we keep track of
whether the search was terminated by the iteration limit.

murchandamus committed:
test: Add coin_grinder_tests

murchandamus reviewed (commented):
Addressed @sipa’s comments, and completely restructured the PR. Added two more tests, and counting of the number of attempts used in CoinGrinder to measure the optimizations.

Ready for review

murchandamus commented:
Mea culpa. Fixed the linter issue.

murchandamus commented:
Added missing test title in overview of CoinGrinder tests

murchandamus reviewed (commented):
Addressed @sipa’s comments

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,168 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+     * use three state transactions to progress from the current selection to the next promising selection:
+     *
+     * - EXPLORE inclusion branch: We do not have sufficient funds yet and the lookahead promises sufficient funds to
+     *                             reach the target. We continue with the direct successor as our `next_utxo` for the
```
```suggestion
     *                             reach the target. We continue with the direct successor of our `next_utxo` for the
```

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,168 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+            max_tx_weight_exceeded = true;
+            should_shift  = true;
+        } else if (curr_amount >= selection_target + change_target) {
+            // Potential solution, adding more weight cannot be better: SHIFT
```
```suggestion
            // Success, adding more weight cannot be better: SHIFT
```

ishaanam reviewed (commented):
Concept ACK

I'm still reviewing this PR, but I have left some initial review comments.

murchandamus reviewed (commented):
Addressed Ishaana’s review

Added a test with a pattern of mixed weight inputs to highlight the improvements from optimizations.

murchandamus committed:
fuzz: Test optimality of CoinGrinder

Co-authored-by: Pieter Wuille <pieter@wuille.net>

murchandamus commented:
I’ve added another fuzz target that validates the output of CoinGrinder against a bruteforce search for the smallest weight input set.

murchandamus commented:
Gawd, SFFO is such pain in the neck.

murchandamus commented:
Fixed bug discovered with the optimality fuzz test: [opt: Skip over barren combinations of tiny UTXOs](https://github.com/bitcoin/bitcoin/pull/27877/commits/80bcee5c09fb68366645d6be62fdeab4bb3ec6f3), incorrectly was set to cut, but this could be premature in case that the last selected UTXO was heavier than later UTXOs in the pool.

murchandamus commented:
CI failure is unrelated, see https://github.com/bitcoin/bitcoin/pull/29243

murchandamus commented:
Rebased on fix for CI issue

murchandamus reviewed (commented):
Addressed most of @sipa’s feedback except for two things I’m still mulling over

murchandamus commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -133,6 +133,82 @@ FUZZ_TARGET(coin_grinder)
+    assert(num_utxos <= max_utxos);
+
+    std::vector<OutputGroup> group_pos;
+    GroupCoins(fuzzed_data_provider, utxo_pool, coin_params, /*positive_only=*/true, group_pos);
```
I cleaned up the Optimality fuzz test a bit:

- I have removed multiple CoinSelectionParams from being fuzzed that are not relevant to testing CoinGrinder’s optimality (SFFO, LTFRE, all change cost related quantities except `min_change_target`)
- I now fuzz `min_change_target`
- I just create up to 16 UTXOs that are each a separate OutputGroup
- I now first run the brute force solution, then set a fuzzed max_weight greater than the best_weight and check that CoinGrinder finds the optimal solution, then set a lower max_weight and check that CoinGrinder provides no solution
- I generate `target` to be below `max_spendable` instead of returning early

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,168 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+    }
+
+    if (best_selection.empty()) {
+        return max_tx_weight_exceeded ? ErrorMaxWeightExceeded() : util::Error();
```
I have decided to defer this for the moment, since I do appreciate the clarity of distinguishing between the `best_selection_weight` and `max_weight`, and I anticipate that I will have to revisit this in the context of https://github.com/bitcoin/bitcoin/pull/29264 soon anyway.

murchandamus commented:
All review feedback is addressed, ready for review.

sipa reviewed (commented):
Concept ACK

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,177 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+            }
+        }
+
+        if (curr_try >= TOTAL_TRIES) {
```
While I agree that CoinGrinder already does work that scales with the size of the UTXO pool, I would argue that CoinGrinder should never need more than 100'000 tries to find _a_ solution. We can only use up to ~1740 inputs in a standard transaction, so if a wallet can’t scrounge up enough funds after traversing 100'000 candidate input sets, I’d say it has other issues.

Happy to change both of these in a follow-up, but wouldn’t consider it necessary for this PR to land.

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,177 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+        if (should_cut) {
+            // Neither adding to the current selection nor exploring the omission branch of the last selected UTXO can
+            // find any solutions. Redirect to exploring Omission branch of the penultimate selected UTXO (i.e. deselect
+            // last selected UTXO, then SHIFT)
```
No, SHIFT is _deselect + explore_ and CUT is _deselect + deselect + explore_, so CUT is _deselect + SHIFT_.

murchandamus replied:
That said, your comment indicates that this could be phrased better, since it appears to be difficult to follow. Will improve.

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,177 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+            // Set `next_utxo` to one after last selected, then deselect last selected UTXO
+            if (curr_selection.empty()) {
+                // Exhausted search space before running into attempt limit
+                result.SetSelectionsEvaluated(curr_try);
```
Yeah, looks correct. I will move `result.SetSelectionsEvaluated(curr_try)` out of the loop next time I’ll push. I know there are some other reviews in flight right now.

murchandamus commented on src/wallet/spend.cpp:
```diff
@@ -709,6 +709,13 @@ util::Result<SelectionResult> ChooseSelectionResult(interfaces::Chain& chain, co
+        if (auto cg_result{CoinGrinder(groups.positive_group, nTargetValue, coin_selection_params.m_min_change_target, max_inputs_weight)}) {
+            cg_result->ComputeAndSetWaste(coin_selection_params.min_viable_change, coin_selection_params.m_cost_of_change, coin_selection_params.m_change_fee);
+            results.push_back(*cg_result);
+        } else append_error(cg_result);
```
Okay, will fix when I push next time.

murchandamus commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -134,6 +135,86 @@ FUZZ_TARGET(coin_grinder)
+    FastRandomContext fast_random_context{ConsumeUInt256(fuzzed_data_provider)};
+    CoinSelectionParams coin_params{fast_random_context};
+    coin_params.m_subtract_fee_outputs = false;
+    coin_params.m_effective_feerate = CFeeRate{ConsumeMoney(fuzzed_data_provider, /*max=*/COIN)};
```
Okay cool, sound great. Will adopt with the next push.

murchandamus reviewed (commented):
Thanks @sipa, good improvements, will amend when the other review in flight comes back.

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -183,6 +200,223 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+            deselect_last();
+            should_shift  = false;
+
+            // After SHIFTing to an omission branch, the `next_utxo` might have the same value and same weight as the
```
Yep, will update.

sr-gi commented on src/wallet/coinselection.cpp:
```diff
@@ -204,13 +204,18 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
-        Assume(utxo.GetSelectionAmount() > 0);
-        total_available += utxo.GetSelectionAmount();
+    size_t i = utxo_pool.size();
+    while (i > 0) {
```
In opt: Track remaining effective_value in lookahead:

I wonder why this gets turned into a descending while loop instead of a descending for loop. Given the total number of iterations is known beforehand a for seems best suited (plus the scope of the counter is reduced to to scope of the loop)

sr-gi commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -1269,6 +1270,33 @@ BOOST_AUTO_TEST_CASE(coin_grinder_tests)
+            }
+            return available_coins;
+        });
+        expected_result.Clear();
```
in test: Exhaust search attempts with tiny UTXOs:

Same as for cases 5) and 6), clearing is not needed

sr-gi reviewed (commented):
Concept ACK.

First pass, I need to dig a bit deeper, but it makes sense to me

murchandamus committed:
opt: Skip heavier UTXOs with same effective value

When two successive UTXOs differ in weight but match in effective value,
we can skip the second if the first is not selected, because all input
sets we can generate by swapping out a lighter UTXOs with a heavier UTXO
of matching effective value would be strictly worse.

murchandamus reviewed (commented):
Addressed feedback by @sipa and @sr-gi.

I also improved the clone skipping by skipping any UTXOs with same effective value and heavier weight that follow an unselected UTXO.

murchandamus commented:
Added a longer textual explanation on how the algorithm can be thought in terms of the search graph it walks.

sr-gi commented on src/wallet/coinselection.cpp:
```diff
@@ -188,6 +188,280 @@ util::Result<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_pool
+ * descendants of the omission branch {A_}, it is unnecessary to evaluate the candidate input set in the omission branch
+ * itself. By skipping evaluation of all nodes on an omission branch we reduce the visited nodes to 15:
+ *
+ *     {} {A} {AB} {ABC} {ABCD} {AB_D} {A_C} {A_CD} {A__D} {_B} {_BC} {_BCD} {_B_D} {__C} {__CD} {__C} {___D} {____}
```
I think this should be 16 though, not 15 as the comment claims:

{} {A} {AB} {ABC} {ABCD} {AB_D} {A_C} {A_CD} {A\__D} {_B} {_BC} {_BCD} {_B_D} {\_\_C} {\_\_CD} ~{\_\_C}~{\_\_\_D} ~{\_\_\_\_}~

murchandamus commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -1090,6 +1090,185 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
+    // 5) Test finding a solution in a UTXO pool with mixed weights
+    // 6) Test that the lightest solution among many clones is found
+
+    FastRandomContext rand;
```
I think that applies to pretty much all coinselection tests, so I’ll leave that for a follow-up at the moment.

murchandamus commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -134,6 +135,84 @@ FUZZ_TARGET(coin_grinder)
+    // Create some coins
+    CAmount max_spendable{0};
+    int next_locktime{0};
+    unsigned max_output_groups = 16;
```
Taken

murchandamus commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -1267,6 +1268,32 @@ BOOST_AUTO_TEST_CASE(coin_grinder_tests)
     }
+
+    {
+        // #################################################################################################################
```
Nope, good point. Folded into `test: Add coin_grinder_tests`

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -448,6 +448,13 @@ util::Result<SelectionResult> CoinGrinder(std::vector<OutputGroup>& utxo_pool, c
                 best_selection_weight = curr_weight;
                 best_selection_amount = curr_amount;
             }
+        } else if (!best_selection.empty() && curr_weight + min_tail_weight[curr_tail] * ((total_target - curr_amount) / utxo_pool[curr_tail].GetSelectionAmount()) > best_selection_weight) {
```
I used the suggestion to upcast and to round up, but I’m leaving the `!best_selection.empty()` in, as we would otherwise not produce a `max_tx_weight_exceeded` error.

murchandamus reviewed (commented):
Responded to feedback from @sipa and @sr-gi

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -473,12 +473,12 @@ util::Result<SelectionResult> CoinGrinder(std::vector<OutputGroup>& utxo_pool, c
+            // ties on the effective value, it _must_ have the same weight (i.e. be a "clone" of the prior UTXO) or a
+            // higher weight. If so, selecting `next_utxo` would produce an equivalent or worse selection as one we
+            // previously evaluated. In that case, increment `next_utxo` until we find a UTXO with a differing amount.
+            while (utxo_pool[next_utxo - 1].GetSelectionAmount() == utxo_pool[next_utxo].GetSelectionAmount()) {
```
I see thanks for the explanation. It seems to me like most of these cases should be caught by the optimization to cut instead of shift when we exceed the weight:

>  opt: Cut if last addition was minimal weight
> 
> In situations where we have UTXO groups of various weight, we can CUT
> rather than SHIFT when we exceeded the max_weight or the best
> selection’s weight while the last step was equal to the minimum weight
> in the lookahead.

It would take more time to determine whether this would make another meaningful impact if implemented, so I would like to defer this to a potential follow-up.

murchandamus reviewed (commented):
Addressed all comments by @sipa and @achow101. Thanks for the review!

achow101 commented:
ACK 13161ecf032b7a850686e5942c12222c8f3d0d52

sipa reviewed (commented):
ACK 13161ecf032b7a850686e5942c12222c8f3d0d52

sr-gi commented:
ACK [13161ec](https://github.com/bitcoin/bitcoin/pull/27877/commits/13161ecf032b7a850686e5942c12222c8f3d0d52)

murchandamus commented on src/wallet/coinselection.cpp:
```diff
@@ -454,6 +455,13 @@ util::Result<SelectionResult> CoinGrinder(std::vector<OutputGroup>& utxo_pool, c
                 best_selection_weight = curr_weight;
                 best_selection_amount = curr_amount;
             }
+        } else if (!best_selection.empty() && curr_weight + int64_t{min_tail_weight[curr_tail]} * std::ceil((total_target - curr_amount) / utxo_pool[curr_tail].GetSelectionAmount()) > best_selection_weight) {
```
Mh, it could also be turned on after max_weight was exceeded the first time. :thinking: Might follow-up on this.

murchandamus commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -77,6 +78,144 @@ static SelectionResult ManualSelection(std::vector<COutput>& utxos, const CAmoun
+        // Only make UTXOs with positive effective value
+        const CAmount input_fee = coin_params.m_effective_feerate.GetFee(n_input_bytes);
+        // Ensure that each UTXO has at least an effective value of 1 sat
+        const CAmount eff_value{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY - max_spendable - max_output_groups + group_pos.size())};
```
I was unable to reproduce the error that @maflcko reported.

Analyzing the line, I don’t see how the content of the parentheses could ever become negative. At most, it would become temporary negative, when `max_spendable` and `max_output_groups` have already been deducted, but `group_pos.size()` has not yet been added.

Would it perhaps help if the addition were put to the front so it never even temporarily goes negative?

```diff
-const CAmount eff_value{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY - max_spendable - max_output_groups + group_pos.size())};
+const CAmount eff_value{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY + group_pos.size() - max_spendable - max_output_groups)};
```

I don’t see how the suggestion @sipa made would ever be triggered.
I would expect that adding an unsigned to a long, would make the result a signed int, so I’m a bit baffled what the issue here is. Would I perhaps need to cast explicitly?

achow101 replied:
> I was unable to reproduce the error that @maflcko reported.

I was able to reproduce if I enabled the integer sanitizer.

> Would it perhaps help if the addition were put to the front so it never even temporarily goes negative?
> 
> ```diff
> -const CAmount eff_value{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY - max_spendable - max_output_groups + group_pos.size())};
> +const CAmount eff_value{fuzzed_data_provider.ConsumeIntegralInRange<CAmount>(1, MAX_MONEY + group_pos.size() - max_spendable - max_output_groups)};
> ```

This change appears to fix this particular issue.

murchandamus replied:
Thanks, I guess I misinterpreted `SUMMARY: UndefinedBehaviorSanitizer: implicit-integer-sign-change ../../src/wallet/test/fuzz/coinselection.cpp:161:89 in` as needing the "undefined" sanitizer, but didn’t realize that there was one for "integer". I was able to reproduce and verify that changing the order of operations mitigates the issue here.
